### DISCUSSION TITLE
메소드 또는 패키지를 삭제한 경우 스냅샷에 반영되는 시점이 궁금합니다.

### DISCUSSION THREAD

**Discussion by kwonho87** [naver/scavenger#123] [Q&A] [answered]
스캐빈저를 통해 미사용으로 판단된 메소드를 삭제하면서 패키지 및 클래스도 삭제 하였습니다.
agent 를 재기동 하면 method 갯수가 그만큼 줄어서 로그에 나오고 있는데,
스냅샷을 생성하면 변경된 method 갯수가 반영이 안되고 있습니다.
스냅샷에서 변경사항이 적용되게 필요한 조치가 어떤게 있을지 궁금합니다.

감사합니다.

**Comment by sohyun-ku** [comment]
@kwonho87 

안녕하세요.
snapshot의 경우 codebase 기준이 아닌 실제 메서드 호출 기록(invocation)을 이용해 생성되고 있습니다.
scavenger는 배치를 통해 더 이상 사용하지 않는 method, invocation를 GC를 하고 있으며,
스냅샷에 변경사항이 적용되는 시점은 더 이상 해당 메서드 호출이 수집되지 않는다는 가정 하에 Collector에서 invocation GC가 일어난 이후(명일 이후)로 예상됩니다.

**Reply by kwonho87** [reply]
답변 주셔서 감사합니다!

collector 에서 invocation GC 가 일어난 이후(명일 이후) 라고 하셨는데,
위 말씀하신 가정 조건에 부합하고 3일이 지났지만 아직 스냅샷에 반영이 안되는 것 같습니다.
혹시 제가 확인해 볼 설정이나 로그 또는 강제로 배치를 수행할 수 있는 방법이 있을까요?

확인 부탁 드립니다.
감사합니다.

**Reply by sohyun-ku** [reply, marked as answer]
jvm이 존재하지 않는 환경의 경우 GC가 일어나지 않기 때문에 아래 두 가지 확인 부탁드립니다.

- snapshot에서 사용하는 모든 환경 애플리케이션에서 메서드가 제거되어있는지 확인
- 해당 환경에 jvm이 존재하는지 확인

GC 가 수행되었을때는 아래와 같은 로그를 확인하실 수 있습니다.
https://github.com/naver/scavenger/blob/a43ca521f25f556f98ca5582bb865daf383b6e58/scavenger-collector/src/main/kotlin/com/navercorp/scavenger/service/GarbageCollectService.kt#L133-L178

**Reply by kwonho87** [reply]
코멘트 주신 내용 참고하여 해결하였습니다.

원인은 초기 scavenger 세팅 시 워크스페이스를 삭제 및 재성성하는 과정에서
jvms 테이블에 동일한 customerId 로 사용하지 않고 있는 uuid 가 존재하게 되었고, 
InvocationImportService에서 invocations 목록 수집이 정상적으로 안된것 같습니다.

jvms 테이블에서 미사용 uuid 삭제 하기 InvocationImportService 에서 invocations 갯수가 업데이트 되는것을 로그로 확인 하였고,
GarbageCollectService 에서 삭제되는것을 확인 하였습니다.

저희 상용 서비스에서 scavenger 를 현재 사용중인데
덕분에 이슈 잘 해결하였습니다.

감사합니다.
